### ISSUE TITLE
QoS Best Effort Reliability Issue

### ISSUE THREAD

**Issue by ciandonovan** [ros2/ros2#1434]
## Bug report

**Required Info:**

- Operating System:
  - Official DockerHub ROS2 images & Ubuntu 22.04, 20.04, & 18.04
- Installation type:
  - binaries
- DDS implementation:
  - FastDDS, CycloneDDS, & RTI Connext

#### Steps to reproduce issue

https://github.com/ciandonovan/ros2_qos_reliability/

tl;dr: using tools such as RViz over WiFi can cause a robot to stall, even when all topics are subscribed using the "best_effort" policy - seemingly it's functionally equivalent to the "reliable" policy

The reproducible demo for convenience uses containers to simulate two remote hosts,
but I've also reproduced this issue outside of containers between two real remote hosts,
and that both the FastDDS and Cyclone Iceyorx shared-memory implementations are also affected by slow remote subscribers.

##### Prerequisites

Webcam at `/dev/video0` and Podman installed.

##### Download and build

```
git clone https://github.com/ciandonovan/ros2_publish_rate.git
cd ros2_qos_reliability
podman build -t $(basename $PWD) .
```

##### Run

`export RMW_IMPLEMENTATION=rmw_fastrtps_cpp` or `export RMW_IMPLEMENTATION=rmw_cyclonedds_cpp`

Simulate a constrained WiFi network on the loopback interface.

`sudo tc qdisc add dev lo root tbf rate 10mbit burst 10mbit latency 50ms`

N.B. use `sudo iftop -i lo` to monitor loopback traffic in real-time.

Run ROS2 camera publisher in one terminal.

`podman run --rm -it --name ros2_publisher --net=host --device=/dev/video0:/dev/video0:rw --env RMW_IMPLEMENTATION ros2_qos_reliability:latest ros2 run v4l2_camera v4l2_camera_node --ros-args --log-level debug`

Run ROS2 camera subscriber in another.

`podman run --rm -it --name ros2_subscriber --net=host --env RMW_IMPLEMENTATION ros2_qos_reliability:latest ros2 topic echo --qos-reliability best_effort /image_raw`

Confirm publisher and subscriber reliability settings

`podman exec -it --env RMW_IMPLEMENTATION ros2_publisher /ros_entrypoint.sh ros2 topic info --verbose /image_raw`

Delete tc qdisc on the loopback when finished.

`sudo tc qdisc del dev lo root`

#### Expected behavior

When running the subscriber with `--qos-reliability reliable` instead, the subscription behaves as expected, the publisher throttles its publishing rate so until the point where all subscribers can receive all published messages given the bandwidth. This can be seen by observing the publisher's debug messages. The "best_effort" subscription should not induce any throttling on the publisher.

#### Actual behavior

The "best_effort" subscription induces throttling on the publisher, the same as what happens with the "reliable" subscription. 

#### Additional information

Mobile robots are very popular in the ROS community, and they necessitate wireless communication for real-time monitoring and debugging. Tools such as RViz are commonly used over a WiFi connection to visualize robot sensor data, path-planning routes, etc. Unlike fixed robots connected via a wired connection where the bandwidth and latency are fairly constant, the bandwidth and latency on a WiFi between the mobile robot and a developer's workstation can vary considerably and often unpredictably as a function of the robot and the developer's relative positions. This shouldn't be a issue, as the developer's view of the robot's own perspective on the world is supplemental to the robot's operation - not critical. This is what the ROS2 topic reliability policies were designed for, a reliable subscription policy that can be used for mission-critical connections, possibly internal to the robot, and a "best_effort" subscription policy that can be used for non-critical monitoring and debugging.

As was demonstrated however, there is seemingly no functional difference between the two policies - both act as a reliable subscriber, meaning that the publisher will throttle its publishing rate if the subscriber can't keep up, which is often the case for high-bandwidth topics over a wireless connection.

This means that for mobile robots, where engineers use tools such as RViz monitoring topics over a WiFi link, should that link degrade, the internal communication on the robot also degrades, even when using shared memory. As such, the behaviour of the robot as a whole under test, and in the field, could be drastically different.

![2023-05-31-Note-23-14](https://github.com/ros2/ros2/assets/94260580/b5bfddbd-80cd-4f4b-ad41-9aa89cc834a6)

For us this often causes our robot to grind to a halt if we're trying to debug the robot's navigation using RViz and we loose line of sight. This could be alleviated by only using ROS2 DDS for internal communication on the host, and a ROS bridge to external devices. But then the benefits of using DDS are lost.

Better WiFi coverage and higher bandwidth APs could also alleviated the issue somewhat, but still offers no guarantees, and in principle should not have an effect at all - such non-determinism is a non-starter in many industrial applications.

Further discussion on ROS Discourse here https://discourse.ros.org/t/ros2-qos-reliability-issue/31700

**Comment by ciandonovan** [comment]
CC: @destogl might have implications for ros2_control, have you come across this issue yourself?

**Comment by ciandonovan** [comment]
@eboasson I also reproduced the issue using FastDDS's shared-memory on the host and a remote subscriber over WiFi. Verified there was no traffic on the loopback interface, shared-memory was definitely being used, and still when a slow remote node subscribed the host subscriber slowed too. Also reproduced with CycloneDDS + Iceoryx shared-memory, and RTI Connext's IPC.

My suspicion is that the way the publisher is implemented round-robin sends to its list of subscribers, always blocks for receipt of a sent message, and so making the slowest common denominator subscriber the limiting factor in the publish rate.

**Comment by ciandonovan** [comment]
@eboasson Yes a bridge looks like the only solution at the moment, we already use Foxglove's WebSocket bridge for WAN traffic to avoid the same issue. The Zenoh bridge looks nice too. It is a shame though to loose DDS's zero-configuration works-out-of-the-box, more work to configure and manage developer machines as well. But can't afford to risk external network factors interfering with the robot's operation, so have to just bite the bullet there!

In the longer term then, the publisher needs to be modified to decouple the flows, a thread for each subscriber or something like that?

**Comment by eboasson** [comment]
> In the longer term then, the publisher needs to be modified to decouple the flows, a thread for each subscriber or something like that?

If each subscriber uses its own topic, you could do that, yes. I do think there is a solution possible at DDS level, at least for this level of complexity. The problem is that I am a bit too busy to promise an experimental implementation sometime soon ...

> It is a shame though to loose DDS's zero-configuration works-out-of-the-box, more work to configure and manage developer machines as well.

Absolutely, but at least with the Zenoh bridge you should remain *really* close to that goal because Zenoh itself does zero-configuration publish/subscribe, integrates with Cyclone to be an ordinary DDS participant for the bridging and uses the discovery information to dynamically build up the bridged data flows. The configuration is needed primarily for limiting the topics or the data rates.

(I am not up-to-date on the Zenoh plugins, there might even be a websocket one!)

**Comment by ciandonovan** [comment]
You mention each subscriber needing its own topic, how about the use case below where there are two subscribers, one reliable for system operation, and one best effort for monitoring?

![qos_reliability](https://github.com/ros2/ros2/assets/94260580/f21355c1-32fd-4e17-8d42-3fee73e38d9a)

According to the ![docs](https://docs.ros.org/en/rolling/Concepts/About-Quality-of-Service-Settings.html) it should be possible, note the section _Compatibility of reliability QoS policies_ in particular. Although I wonder if it ever got implemented properly in the first place, maybe was overlooked? I've tested all the way back to Foxy anyway.

The 10 mbps example is a contrived limit, I wouldn't actually try stream much over a connection like that, but it does work to show clearly the detrimental effect a "best_effort" subscriber can have. Often for us it manifests as jitter on the reliable subscribers, which isn't ideal for building deterministic systems.

From digging into Cyclone a bit, I see that the `dds_qset_reliability()` function is being used to set the policy on the DDS level, but I'm guessing that the ROS2 publisher a layer above isn't feeding it data continuously because it's blocking elsewhere?

If you know where in the various repos the handoff from ROS2 to the DSS is, or where the publishing side is handled, I can start to have a look myself.

**Comment by eboasson** [comment]
> According to the docs it should be possible, note the section Compatibility of reliability QoS policies in particular. Although I wonder if it ever got implemented properly in the first place, maybe was overlooked? I've tested all the way back to Foxy anyway. 

Clever 🙂 The QoS policy matching has always been part of DDS. I think it was an unwise choice of the spec writers, but given that it exists, yes, you can indeed build on it.

It is unfortunate that ROS doesn't expose the DDS partitions, those are actually the feature really meant to be used for this type of thing. If you are into hacking ROS to suit your needs, it would actually be a straightforward bit of drudgery to add that.

It kinda goes rclcpp -> rcl -> rmw -> rmw implementation (e.g., rmw_cyclonedds_cpp). The ROS 2 QoS profiles are defined in https://github.com/ros2/rmw/blob/rolling/rmw/include/rmw/qos_profiles.h, the translation of these to DDS happen inside the RMW implementation, for Cyclone in: https://github.com/ros2/rmw_cyclonedds/blob/ee3485a5d34cd052c55d43ce24379b68650f3317/rmw_cyclonedds_cpp/src/rmw_node.cpp#L2193

All the QoS stuff happens when the publisher/subscriber is being created, after that it is just data flowing through the system.

There is a snag though, if you'd want to use partitions: in DDS partitions are bound to "publisher" and "subscriber" objects, and the ROS 2 "publisher" and "subscriber" map to "writer" and "reader" in DDS. So if you want to use multiple partitions, you'd need multiple publishers/subscribers in DDS.

Fortunately, with Cyclone that's easy enough: unlike the DDS spec it allows you to pass a DDS participant as the owner of the reader/writer, in which case it will spontaneously create a DDS subscriber/publisher as well using the same QoS object you pass `dds_create_reader`, &c. Thanks to how Cyclone represents its QoS's that means you if you set the partition QoS in the QoS object you pass to `dds_create_reader` and the owner is the participant, a subscriber gets created using that partition.

And, last but not least, if you delete a reader created like that, it also deletes that subscriber. Yay!

So all it takes is modifying https://github.com/ros2/rmw_cyclonedds/blob/ee3485a5d34cd052c55d43ce24379b68650f3317/rmw_cyclonedds_cpp/src/rmw_node.cpp#L3036 to pass `dds_ppant` instead of `dds_sub` (and likewise for the call to `dds_create_writer`) 😀

**Comment by fujitatomoya** [comment]
@eboasson @ciandonovan really appreciate for sharing experience and information, that is really good information for us and ROS 2 users. (and thanks for being responsive too 👍 ) I think this information should be described in documentation as well in ROS 2 doc? and for long-term we need to address this problem as ROS 2 since this is not good user experience considering the use case of ROS 2 distributed system.

> There is a snag though, if you'd want to use partitions: in DDS partitions are bound to "publisher" and "subscriber" objects, and the ROS 2 "publisher" and "subscriber" map to "writer" and "reader" in DDS. So if you want to use multiple partitions, you'd need multiple publishers/subscribers in DDS.

I see.

**Comment by ciandonovan** [comment]
@eboasson thanks so much for that excellent detailed description of the architecture and how the different layers interact, it's a great starting point! I'm a CycloneDDS user on Humble myself so very motivated to get hacking there :)

If it is related to how ROS2 utilizes the underlying DDS partitions, then work might have to be done with Fast-DDS/rmw_fastrtps as well, especially considering it's the default DDS vendor at the moment - agree strongly with @fujitatomoya about the importance of a good user experience here.

CC: @JaimeMartin @EduPonz

**Comment by EduPonz** [comment]
Hi @ciandonovan,

Thanks for the report and interesting discussion! Since you guys already went over some of the options, I'll try to summarize possibilities when using Fast DDS:

1. If the issue comes from the writer blocking on the UDP transport, we'd suggest to configure the transport in non-blocking mode. For that, you'd need to:
    1. Create a Fast DDS XML configuration profile which disables the builtin transports and creates two user transports, one UDPv4 and one SHM (similar to the default ones). [Here](https://docs.vulcanexus.org/en/latest/rst/tutorials/core/deployment/transport/custom_transports.html) you can read a tutorial on how to do so, and [here](https://fast-dds.docs.eprosima.com/en/latest/fastdds/xml_configuration/transports.html#transport-descriptors) you can read the documentation regarding Fast DDS transport configuration via XML.
    1. The important bit is to add the `<non_blocking_send>true</non_blocking_send>` tag to the UDPv4 transport descriptor. This will make the call to write to the socket non-blocking. Doing so may imply that some messages will never be sent to the best effort readers, but from what I understand that's OK for you, as it will prevent the writing to the socket to block the delivery to local readers.
1. Adding a local proxy is also possible with Fast DDS, in this case without the need of switching protocols, so you'd only need a proxy on the sending machine, but that will be transparent to the receivers. This can be done with [DDS Router](https://eprosima-dds-router.readthedocs.io/en/latest/index.html), more particularly using it to [change domains](https://docs.vulcanexus.org/en/latest/rst/tutorials/cloud/change_domain/change_domain.html). This configures the DDS Router as a repeater of one DDS domain into another, so you'd simply need to run the remote subscribers into that other domain. Mind that with this option it is possible to curate the list of topics that you want to repeat to the other domain, so you control which information leaves the machine.
1. The alternative regarding partitions can also be explored, as in Fast DDS it is possible to [configure partitions at the DataWriter and DataReader level](https://fast-dds.docs.eprosima.com/en/latest/fastdds/property_policies/non_consolidated_qos.html#endpoint-partitions). This will however entail the application to have two ROS 2 publications on the same topic but different partitions (one for local and the other for remote comms).

As complementary information, I'm leaving [here](https://docs.vulcanexus.org/en/latest/rst/tutorials/core/qos/xml_profiles/xml_profiles.html) documentation on how to configure Fast DDS using XML profiles in ROS 2.

**Comment by ros-discourse** [comment]
This issue has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros2-qos-reliability-issue/31700/23

**Comment by EduPonz** [comment]
> is this the same setting with https://github.com/ros2/rmw_fastrtps#change-publication-mode?

No, they are different settings. Publication mode allows for controlling whether the data is written into the transport within the same thread that calls publish (synchronous) or from a separate thread internal to Fast DDS (asynchronous). The non blocking send flag is used on the socket based transports to control whether the `send` operation should block until the data can be written into the corresponding socket (regardless of the thread in which this operation takes place).

They confusion may come from the default configuration in ROS 2, i.e. synchronous publication and blocking send. In that scenario, were the writing into the socket block for whatever reason, that will cause the application thread calling `publish` to also block. This is precisely why the default mode used to be asynchronous in ROS 2 (with Fast DDS), although that changed after Galactic, in which the default middleware was CycloneDDS , which at least at the time only supported synchronous publishing. On the other hand, synchronous publishing yields better latency performance results, and that's why after Galactic it was decided as the default settings as far as I know.
